File: web/screens/takeoff.js
// The finishing take-off -- where the rupees are.
//
// This replaces `Internal Finishes Flats`: 1,451 hand-written rows and 9,472
// formulas, one block per room, each re-anchored to the rate list by a
// hand-counted row offset. Here every line is a fold over the room's own
// finishes, and every total is a filter rather than a SUM over a range.

import { api, fmt, route } from '../app.js';
import { createGrid } from '../grid.js';
import { escapeHtml, showContributors, showDerivation } from '../panel.js';

const STATUS = {
  priced: '', no_rate: 'bad', no_rule: 'warn', error: 'bad',
};

route('/takeoff', async (main) => {
  main.innerHTML = '<div class="loading">Measuring every room and pricing it…</div>';
  const t = await api.get('/takeoff');

  main.innerHTML = `
    <div class="screen-head">
      <h1>Finishing take-off</h1>
      <p>Every finish, in every room, of every unit type — measured, deducted, priced and
         multiplied by how many of that unit the building has. Nothing here was typed.</p>
    </div>

    <div class="tile-row">
      <div class="tile"><div class="k">Finishing cost</div>
        <div class="v">${fmt.money(t.total)}</div>
        <div class="s">${fmt.int(t.priced_count)} priced lines</div></div>
      <div class="tile"><div class="k">Take-off lines</div>
        <div class="v">${fmt.int(t.line_count)}</div>
        <div class="s">was 1,451 rows in the workbook</div></div>
      <div class="tile"><div class="k">Measured, unpriced</div>
        <div class="v" style="color:${t.unpriced.length ? 'var(--bad)' : 'inherit'}">
          ${fmt.int(t.unpriced.length)}</div>
        <div class="s">${t.unpriced.length ? 'work showing nothing' : 'none'}</div></div>
      <div class="tile"><div class="k">Finish types</div>
        <div class="v">${fmt.int(t.by_finish.length)}</div>
        <div class="s">across ${t.by_unit_type.length} unit types</div></div>
    </div>

    ${t.unpriced.length ? `
    <div class="card" style="margin-top:16px;border-color:#f3caca">
      <h2>Measured but not priced
        <span class="sub">quantity known, rate missing — this is not zero-cost work</span></h2>
      <div id="unpriced"></div>
    </div>` : ''}

    <div class="card" style="margin-top:16px">
      <h2>By finish <span class="sub">click a blended rate to see what it is made of</span></h2>
      <div id="byFinish"></div>
    </div>
    <div class="card">
      <h2>By unit type</h2>
      <div id="byUnit"></div>
    </div>
    <div class="card">
      <h2>All lines <span class="sub">${fmt.int(t.line_count)} — filter to narrow</span></h2>
      <div class="card-body" style="padding-bottom:0">
        <input class="text-input" id="q" placeholder="Filter by unit type, room or finish…"
               style="min-width:320px">
        <span class="muted" id="count"></span>
      </div>
      <div id="lines"></div>
    </div>`;

  if (t.unpriced.length) {
    createGrid(document.getElementById('unpriced'), {
      columns: [
        { key: 'unit_type', label: 'Unit type', kind: 'label', width: '140px' },
        { key: 'room', label: 'Room', kind: 'label', width: '170px' },
        { key: 'finish', label: 'Finish', kind: 'note', width: '160px', align: 'left' },
        { key: 'total_qty', label: 'Quantity', kind: 'derived', dp: 2, width: '110px' },
        { key: 'unit', label: 'Unit', kind: 'note', width: '60px', align: 'left' },
        { key: 'message', label: 'Why', kind: 'note', width: '380px', align: 'left',
          render: v => `<span class="muted">${escapeHtml(v)}</span>` },
      ],
      rows: t.unpriced,
      onDerivedClick: async (row, col) => {
        if (col.key !== 'total_qty') return false;
        const full = await working(row);
        showDerivation(`${row.room} — ${row.finish}`, row.total_qty,
          full.gross_derivation, {
            unit: row.unit,
            extra: `<div class="deriv-note">${escapeHtml(row.message)}</div>
              <div class="deriv-note">This quantity is real. It is the amount
                beside it that is missing, and the two are not the same
                thing.</div>`,
          });
        return true;
      },
    });
  }

  createGrid(document.getElementById('byFinish'), {
    columns: [
      { key: 'label', label: 'Finish', kind: 'label', width: '230px' },
      { key: 'quantity', label: 'Quantity', kind: 'derived', dp: 2, width: '124px' },
      { key: 'unit', label: 'Unit', kind: 'note', width: '60px', align: 'left' },
      { key: 'blended_rate', label: 'Blended rate', kind: 'derived', dp: 2, width: '116px',
        title: 'Amount ÷ quantity. The workbook prices its cost sheet on exactly this '
             + 'figure — a weighted average that exists in no rate list.' },
      { key: 'lines', label: 'Lines', kind: 'derived', dp: 0, width: '68px' },
      { key: 'amount', label: 'Amount', kind: 'derived', width: '134px', total: true,
        render: v => fmt.money(v) },
    ],
    // Slots that no room measures would be a page of zeros. They are counted
    // below rather than listed, because a zero with no quantity behind it says
    // nothing -- and a zero *with* a quantity is the unpriced table above.
    rows: t.by_finish.filter(g => g.quantity || g.amount),
    rowKey: r => r.key,
    onDerivedClick: (row, col) => groupWorking(row, col, 'unit type'),
    footer: rows => ['<strong>Total</strong>', '', '', '', '',
                     `<strong>${fmt.money(rows.reduce((a, r) => a + r.amount, 0))}</strong>`],
  });

  const empty = t.by_finish.filter(g => !g.quantity && !g.amount);
  if (empty.length) {
    document.getElementById('byFinish').insertAdjacentHTML('beforeend', `
      <div class="card-body muted" style="border-top:1px solid var(--line)">
        ${empty.length} finish slot${empty.length === 1 ? '' : 's'} exist in the rate library
        but no room measures ${empty.length === 1 ? 'it' : 'them'}:
        ${empty.map(g => escapeHtml(g.label)).join(' · ')}.
        Look for near-duplicates here — a slot spelled two ways splits one finish in two.
      </div>`);
  }

  createGrid(document.getElementById('byUnit'), {
    columns: [
      { key: 'label', label: 'Unit type', kind: 'label', width: '200px' },
      { key: 'lines', label: 'Lines', kind: 'derived', dp: 0, width: '76px' },
      { key: 'unpriced', label: 'Unpriced', kind: 'derived', dp: 0, width: '84px' },
      { key: 'amount', label: 'Amount', kind: 'derived', width: '150px', total: true,
        render: v => fmt.money(v) },
    ],
    rows: t.by_unit_type.filter(g => g.amount > 0 || g.unpriced),
    rowKey: r => r.key,
    onDerivedClick: (row, col) => groupWorking(row, col, 'finish'),
  });

  /** A group row on either fold: what it is made of, or why a count is what it is. */
  function groupWorking(row, col, brokenDownBy) {
    const rows = t.contributors[row.key] || [];
    if (col.key === 'blended_rate') {
      showContributors(row, rows, {
        title: 'blended rate',
        extra: `<div class="deriv-expr">${fmt.money(row.amount)} ÷ ${
          fmt.n(row.quantity, 2)} ${escapeHtml(row.unit || '')} = ₹${
          fmt.n(row.blended_rate, 2)} per ${escapeHtml(row.unit || '')}</div>`,
        note: `A weighted average across every room that carries this finish —
          it exists in no rate list, and the workbook prices its cost sheet on
          exactly this figure. Rooms priced at different rates pull it about,
          which is why the rows above are worth reading.`,
      });
      return true;
    }
    if (col.key === 'unpriced') {
      showContributors(row, rows, {
        title: `${row.unpriced} unpriced line${row.unpriced === 1 ? '' : 's'}`,
        note: `Measured here and reaching no rate. Their quantity is real and
          their amount is missing rather than zero — that is the C-11 failure
          this platform exists to stop, so they are listed in the table at the
          top of this screen rather than quietly dropped.`,
      });
      return true;
    }
    if (['amount', 'quantity', 'lines'].includes(col.key)) {
      showContributors(row, rows, {
        note: `Broken down by ${brokenDownBy}. Each row is already multiplied by
          how many units the building has. This is a filter over the same
          take-off lines the table below uses, so the two cannot disagree.`,
      });
      return true;
    }
    return false;
  }

  // -- every line, filterable ---------------------------------------------
  function drawLines() {
    const q = document.getElementById('q').value.trim().toLowerCase();
    const rows = t.lines.filter(l => !q ||
      `${l.unit_type} ${l.room} ${l.finish} ${l.rate_description}`.toLowerCase().includes(q));
    document.getElementById('count').textContent =
      `  ${rows.length} of ${t.lines.length} lines`;

    const host = document.getElementById('lines');
    host.innerHTML = '';
    createGrid(host, {
      columns: [
        { key: 'unit_type', label: 'Unit type', kind: 'label', width: '130px' },
        { key: 'room', label: 'Room', kind: 'label', width: '160px' },
        { key: 'finish', label: 'Finish', kind: 'note', width: '150px', align: 'left' },
        { key: 'net', label: 'Net / unit', kind: 'derived', dp: 3, width: '96px' },
        { key: 'unit', label: 'Unit', kind: 'note', width: '56px', align: 'left' },
        { key: 'unit_count', label: '× units', kind: 'derived', dp: 0, width: '70px' },
        { key: 'total_qty', label: 'Total qty', kind: 'derived', dp: 2, width: '104px' },
        { key: 'rate', label: 'Rate', kind: 'derived', dp: 2, width: '96px',
          flagMissing: true,
          render: (v, r) => v === null
            ? `<span class="tag ${STATUS[r.status] || 'warn'}">no rate</span>`
            : `₹${fmt.n(v, 2)}` },
        { key: 'total_amount', label: 'Amount', kind: 'derived', width: '124px',
          total: true, render: (v, r) => r.status === 'priced' ? fmt.money(v) : '—' },
      ],
      rows: rows.slice(0, 400),
      emptyMessage: 'Nothing matches that filter.',
      onDerivedClick: async (row, col) => {
        // The working is fetched for the one figure clicked. Shipping all three
        // derivations on every line made this payload 2 MB, 54% of it panels
        // nobody had opened.
        const full = await working(row);
        if (col.key === 'total_qty' || col.key === 'net') {
          showDerivation(`${row.room} — ${row.finish}`, row.net, full.gross_derivation, {
            unit: row.unit,
            extra: full.deduction_derivation ? `
              <h4 class="deriv-h">Less, for this room's own openings</h4>
              <div class="deriv-expr">${escapeHtml(full.deduction_derivation.expression)}</div>
              ${(full.deduction_derivation.inputs || []).map(i => `
                <div class="deriv-input"><div class="n">${escapeHtml(i.name)}</div>
                  <div class="v">−${fmt.n(i.value, 3)}</div></div>`).join('')}
              ${full.deduction_derivation.note
                ? `<div class="deriv-note">${escapeHtml(full.deduction_derivation.note)}</div>` : ''}
              <h4 class="deriv-h">Then</h4>
              <div class="deriv-expr">${fmt.n(row.net, 3)} ${escapeHtml(row.unit)}
                × ${row.unit_count} units = ${fmt.n(row.total_qty, 2)}</div>` : '',
          });
        } else if (col.key === 'unit_count') {
          showDerivation(`${row.unit_type} — units this line covers`,
            row.unit_count, null, {
              format: v => `${v} unit${v === 1 ? '' : 's'}`,
              extra: `<div class="deriv-note">Counted from the floor matrix in
                Room Config, never typed.${row.floor_scope
                  ? ` This line covers ${escapeHtml(row.floor_scope)}${
                      row.floor_height_m
                        ? `, measured at ${fmt.n(row.floor_height_m, 2)} m
                           floor-to-floor` : ''}: a unit type on floors of more
                     than one height is measured once per height rather than
                     averaged into one wall.` : ''}</div>`,
            });
        } else if (col.key === 'rate' || col.key === 'total_amount') {
          showDerivation(row.rate_description || row.finish, row.rate, full.rate_derivation, {
            unit: `per ${row.unit}`,
            format: v => v === null ? '— no rate' : `₹${fmt.n(v, 4)}`,
            extra: row.status === 'priced' ? `
              <h4 class="deriv-h">Amount</h4>
              <div class="deriv-expr">${fmt.n(row.total_qty, 2)} ${escapeHtml(row.unit)}
                × ₹${fmt.n(row.rate, 2)} = ${fmt.money(row.total_amount)}</div>`
              : `<div class="deriv-note">${escapeHtml(row.message)}</div>`,
          });
        } else {
          return false;
        }
        return true;
      },
    });
  }
  document.getElementById('q').addEventListener('input', drawLines);
  drawLines();
});


// One line's working, fetched on demand and remembered for the session.
const workingCache = new Map();

async function working(row) {
  const key = `${row.room_id}|${row.finish_slot_id}|${row.floor_height_m ?? ''}`;
  if (!workingCache.has(key)) {
    const query = new URLSearchParams({
      room_id: row.room_id, finish_slot_id: row.finish_slot_id,
    });
    if (row.unit_type_id) query.set('unit_type_id', row.unit_type_id);
    if (row.floor_height_m != null) query.set('floor_height_m', row.floor_height_m);
    try {
      workingCache.set(key, await api.get(`/takeoff/derivation?${query}`));
    } catch {
      workingCache.set(key, row);        // fall back to whatever the row holds
    }
  }
  return workingCache.get(key);
}

File: packages/engine/qs_engine/rules/takeoff.py
"""The finishing take-off: quantities meet rates, and rupees appear.

This replaces ``Internal Finishes Flats`` -- 1,451 hand-written take-off rows
and 9,472 formulas, one block per room, each block re-anchored to the rate list
by a hand-counted row offset.

Here it is a fold. For every unit type, every room in it, and every finish that
applies to that room's type: compute the quantity with its deduction, resolve
the rate by identity, multiply by how many of that unit the building has. No
new arithmetic is invented -- this composes ``compute_room_quantity``,
``effective_rate`` and ``amount``, each already tested against the workbook.

Three things the workbook could not do fall out of it:

* A finish that has a quantity and no rate is a **priced-at-nothing** line, not
  a zero. ``Cost Sheet Tower!I99`` shows Rs 0 against 4,508.24 sq.m of measured
  false ceiling while the cell beside it works out Rs 65.5 lakh (C-11).
* A line whose quantity and rate disagree on dimension **raises** rather than
  computing (C-35).
* Every line carries its provenance, so any figure opens into the working.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..model import ProjectModel, UnitTypeRoom
from ..params import ParameterSet
from ..provenance import Derived
from ..rules.rate_buildup import RateBuildupError, effective_rate
from ..rules.room_qty import (HEIGHT_DEPENDENT_RULES, NegativeNetQuantityError,
                              QtyRuleError, compute_room_quantity)
from ..units import Quantity, Rate, UnitConverter, UnitMismatchError, amount


class LineStatus:
    PRICED = "priced"
    NO_RATE = "no_rate"
    NO_RULE = "no_rule"
    ERROR = "error"


@dataclass
class TakeoffLine:
    """One finish, in one room, of one unit type."""

    unit_type_id: str
    unit_type_code: str
    room_id: str
    room_label: str
    room_type_id: str
    room_type_name: str
    finish_slot_id: str
    finish_name: str
    qty_rule: str
    unit: str
    gross: float
    deduction: float
    net: float
    unit_count: int
    total_qty: float
    rate_item_id: str | None
    rate_description: str
    rate: float | None
    amount_per_unit: float
    total_amount: float
    status: str
    message: str = ""
    #: The floor-to-floor height this line was measured against, and the floors
    #: it covers.  A unit type spanning two heights yields two wall lines.
    floor_height_m: float | None = None
    floor_scope: str = ""
    #: Kept so the derivation panel can show the working behind any figure.
    gross_derivation: Derived | None = field(default=None, repr=False)
    deduction_derivation: Derived | None = field(default=None, repr=False)
    rate_derivation: Derived | None = field(default=None, repr=False)

    @property
    def is_priced(self) -> bool:
        return self.status == LineStatus.PRICED


def _placements(model: ProjectModel, unit, rule: str | None
                ) -> list[tuple[float | None, int, str]]:
    """How many lines this unit type needs for this rule, and at what heights.

    A height-driven rule on a unit type that spans two floor heights becomes two
    lines -- an Office is 2.9 m on the podiums and 4.2 m on the ground floor, and
    averaging those would be inventing a building that does not exist.  Every
    other rule collapses back to one line, so flooring is not split into
    fragments that only get re-summed.

    The counts always sum to ``unit_count``, whichever branch is taken.
    """
    placements = model.height_placements(unit.id)
    if not placements:
        return [(None, model.unit_count(unit.id), "")]

    if rule in HEIGHT_DEPENDENT_RULES and len(placements) > 1:
        return [(p.height_m, p.count, _scope(p)) for p in placements]

    heights = {p.height_m for p in placements}
    return [(placements[0].height_m if len(heights) == 1 else None,
             sum(p.count for p in placements), "")]


def _scope(placement) -> str:
    """A short label naming the floors a split line covers."""
    names = placement.floors
    if not names:
        return ""
    if len(names) <= 2:
        return ", ".join(names)
    return f"{names[0]} + {len(names) - 1} more"


def compute_takeoff(model: ProjectModel, params: ParameterSet,
                    unit_type_id: str | None = None) -> list[TakeoffLine]:
    """Every priced finish in the project, or in one unit type."""
    converter = UnitConverter(params["factor_sqm_to_sqft"],
                              params["factor_ft_to_rm"])
    slots = {s.id: s for s in model.finish_slots}
    rate_items = {r.id: r for r in model.rate_items}

    lines: list[TakeoffLine] = []
    unit_types = ([model.unit_type(unit_type_id)] if unit_type_id
                  else sorted(model.unit_types, key=lambda u: u.seq))

    for unit in unit_types:
        for room in model.rooms_of(unit.id):
            for spec in model.finish_spec_for(room.room_type_id):
                slot = slots.get(spec.finish_slot_id)
                if slot is None:
                    continue
                rule = spec.qty_rule or slot.qty_rule
                for height, count, scope in _placements(model, unit, rule):
                    lines.append(_line(model, params, converter, unit, count,
                                       room, spec, slot, rule, rate_items,
                                       height, scope))
    return lines


def _line(model, params, converter, unit, count, room: UnitTypeRoom,
          spec, slot, rule, rate_items, floor_height_m=None,
          floor_scope="") -> TakeoffLine:
    base = dict(
        unit_type_id=unit.id, unit_type_code=unit.code,
        room_id=room.id, room_label=room.label, room_type_id=room.room_type_id,
        room_type_name=model.room_type(room.room_type_id).name,
        finish_slot_id=slot.id, finish_name=slot.name, qty_rule=rule or "",
        unit="", gross=0.0, deduction=0.0, net=0.0,
        unit_count=count, total_qty=0.0,
        rate_item_id=spec.rate_item_id, rate_description="", rate=None,
        amount_per_unit=0.0, total_amount=0.0, status=LineStatus.ERROR,
        floor_height_m=floor_height_m, floor_scope=floor_scope,
    )

    if not rule:
        return TakeoffLine(**{**base, "status": LineStatus.NO_RULE,
                              "message": f"{slot.name} has no quantity rule"})

    try:
        quantity = compute_room_quantity(room, rule, model, params, converter,
                                         floor_height_m)
    except (QtyRuleError, NegativeNetQuantityError, UnitMismatchError) as exc:
        return TakeoffLine(**{**base, "status": LineStatus.ERROR,
                              "message": str(exc)})

    net: Quantity = quantity.net
    total_qty = net.value * room.count_per_unit * count
    base.update(unit=net.unit.code, gross=quantity.gross.value,
                deduction=quantity.deduction.value, net=net.value,
                total_qty=total_qty,
                gross_derivation=quantity.gross_derivation,
                deduction_derivation=quantity.deduction_derivation)

    item = rate_items.get(spec.rate_item_id) if spec.rate_item_id else None
    if item is None:
        return TakeoffLine(**{**base, "status": LineStatus.NO_RATE,
                              "message": f"{slot.name} in {room.label} is "
                                         f"measured but has no rate"})
    base["rate_description"] = item.description

    try:
        rate_derived = effective_rate(item, model, params)
    except RateBuildupError as exc:
        return TakeoffLine(**{**base, "status": LineStatus.NO_RATE,
                              "message": str(exc)})

    revision = model.current_revision(item.id)
    if revision is not None and not revision.is_priced:
        return TakeoffLine(**{
            **base, "status": LineStatus.NO_RATE, "rate": None,
            "rate_derivation": rate_derived,
            "message": f"{item.description!r} has no price. This is measured "
                       f"work showing nothing, not work that costs nothing."})

    try:
        per_unit = amount(net, Rate.of(rate_derived.value, item.unit), converter)
    except UnitMismatchError as exc:
        return TakeoffLine(**{**base, "status": LineStatus.ERROR,
                              "rate": rate_derived.value,
                              "message": str(exc)})

    return TakeoffLine(**{
        **base, "rate": rate_derived.value, "rate_derivation": rate_derived,
        "amount_per_unit": per_unit * room.count_per_unit,
        "total_amount": per_unit * room.count_per_unit * count,
        "status": LineStatus.PRICED})


# --------------------------------------------------------------------------
# Aggregation -- filters, never ranges
# --------------------------------------------------------------------------

@dataclass
class Group:
    key: str
    label: str
    unit: str
    quantity: float
    amount: float
    lines: int
    unpriced: int = 0
    #: The same quantity in square feet, for areas only.  The conversion uses
    #: the project's own ``factor_sqm_to_sqft`` (10.764), not the exact 10.7639,
    #: so the figure agrees with the workbook's.
    quantity_sqft: float | None = None

    @property
    def rate_per_sqft(self) -> float | None:
        """Amount over square feet.  Derived, and labelled as derived.

        A QS reads areas in square feet and rates per square foot; the money is
        still computed from the square-metre pair, so this is a presentation of
        the same number and never an input to it.
        """
        if not self.quantity_sqft:
            return None
        return self.amount / self.quantity_sqft

    @property
    def blended_rate(self) -> float | None:
        """Amount over quantity.

        The workbook's cost sheet is priced on exactly this figure --
        ``Internal Finishes Flats!E1998 = IF(D1998=0,0,F1998/D1998)`` -- a
        weighted average that exists nowhere in the rate list. It is shown
        here as what it is, beside the master rates, rather than standing in
        for them (C-6).
        """
        return self.amount / self.quantity if self.quantity else None


def by_finish(lines: list[TakeoffLine],
              params: ParameterSet | None = None) -> list[Group]:
    """Totals per finish slot, across every unit type in the project.

    This is the whole building's flooring, the whole building's skirting -- the
    view the workbook never had, because its totals were per block.
    """
    return _group(lines, lambda l: (l.finish_slot_id, l.finish_name), params)


def by_unit_type(lines: list[TakeoffLine],
                 params: ParameterSet | None = None) -> list[Group]:
    return _group(lines, lambda l: (l.unit_type_id, l.unit_type_code), params)


def by_room(lines: list[TakeoffLine],
            params: ParameterSet | None = None) -> list[Group]:
    return _group(lines, lambda l: (l.room_id, l.room_label), params)


def by_room_type(lines: list[TakeoffLine],
                 params: ParameterSet | None = None) -> list[Group]:
    """Totals per room type -- every Bedroom in the building, every Toilet."""
    return _group(lines, lambda l: (l.room_type_id, l.room_type_name), params)


def by_finish_and_room_type(lines: list[TakeoffLine],
                            params: ParameterSet | None = None) -> list[Group]:
    """The matrix cell: one finish, in one kind of room, building-wide.

    "Total flooring area in toilets" is one of these, and it is a filter over
    the same lines as every other total, so it cannot disagree with them.
    """
    return _group(
        lines,
        lambda l: (f"{l.finish_slot_id}|{l.room_type_id}",
                   f"{l.finish_name} — {l.room_type_name}"),
        params)


def _group(lines: list[TakeoffLine], key,
           params: ParameterSet | None = None) -> list[Group]:
    """Fold lines into totals, one bucket per key.

    Quantities only add up within one unit, so a bucket holding both square
    metres and running metres reports the money and leaves the quantity blank
    rather than summing the two.  A line that failed to measure carries no unit
    at all, and is counted as unpriced without being allowed to blank out the
    unit of the lines that did measure.
    """
    sqft = params["factor_sqm_to_sqft"] if params is not None else None
    buckets: dict[str, list[TakeoffLine]] = {}
    labels: dict[str, str] = {}
    for line in lines:
        group_key, label = key(line)
        buckets.setdefault(group_key, []).append(line)
        labels.setdefault(group_key, label)

    groups: list[Group] = []
    for group_key, rows in buckets.items():
        units = {r.unit for r in rows if r.unit}
        unit = units.pop() if len(units) == 1 else ""
        quantity = sum(r.total_qty for r in rows if r.unit == unit) if unit else 0.0
        group = Group(
            key=group_key, label=labels[group_key], unit=unit,
            quantity=quantity,
            amount=sum(r.total_amount for r in rows),
            lines=len(rows),
            unpriced=sum(1 for r in rows if not r.is_priced),
        )
        if sqft and unit == "SQM":
            group.quantity_sqft = quantity * sqft
        groups.append(group)
    return sorted(groups, key=lambda g: -g.amount)


def total_amount(lines: list[TakeoffLine]) -> float:
    """The project's finishing cost. A filter, never ``SUM(I39:I99)``."""
    return sum(l.total_amount for l in lines if l.is_priced)


def unpriced(lines: list[TakeoffLine]) -> list[TakeoffLine]:
    """Measured work that reaches no total -- the C-11 class, made visible."""
    return [l for l in lines if l.status == LineStatus.NO_RATE and l.total_qty]
